### DISCUSSION TITLE
preventing sql injections

### DISCUSSION THREAD

**Discussion by I-m-good-man** [isoos/postgresql-dart#329] [Q&A] [answered]
I've read the documentation, but I still don't understand if your library provides protection against sql injections. If there is such a possibility, which method should I use? In which cases protection against sql injections is not implemented

**Comment by isoos** [comment, marked as answer]
@I-m-good-man: SQL injection happens when you are building a query string that includes the parameters inlined. If you want to do that, nobody can prevent it, however, it you are using the `query` method with the `parameters`, these values will not be inlined, instead sent as a separate part of the query. It prevents injection attacks, and also optimizes the query execution, as it spares the server a bit of time with a more efficient protocol.

There are multiple ways to specify parameter locations inside the query, you can see examples using `@name` or `$2`, but also `?` could be used (each with different `Sql` initializer). The `parameters` object should be a `Map` for named parameters or `List` for indexed one. The type safety and specificity can be enhanced by using `TypedValue` instead of raw Dart types. (it will use the more efficient binary encoding this way).

**Reply by I-m-good-man** [reply]
> If you want to do that, nobody can prevent it, however, it you are using the query method with the parameters, these values will not be inlined, instead sent as a separate part of the query.

in this case, what does a query to a database in sql look like? how can parameter values be passed separately from the query string?

**Reply by isoos** [reply]
A few examples:

This query has a single positional parameter:
https://github.com/isoos/postgresql-dart/blob/master/example/example.dart#L45-L48

This query has a single named parameter:
https://github.com/isoos/postgresql-dart/blob/master/example/example.dart#L52-L55

Multiple named parameters:
https://github.com/isoos/postgresql-dart/blob/master/test/query_test.dart#L103-L156

Internally every such parameter gets converted to a query string with `$1`, `$2`, ... as parameter placeholders, and the values will be converted to a list, having the appropriate value in the given position. The binary protocol then sends the query and the parameter values (encoded efficiently) separately.

**Reply by I-m-good-man** [reply]
> Internally every such parameter gets converted to a query string with $1, $2, ... as parameter placeholders

do I understand correctly, when the Connection.execute or Connection.prepared method is called, does the library under the hood create a prepared statement for these queries? 
in the case of Connection.execute, this prepared statement is called immediately and released, and in the case of Connection.prepare, the prepared statement exists until we release it ourselves using Statement.dispose or until we close the connection

**Reply by isoos** [reply]
That is correct.

**Reply by I-m-good-man** [reply]
Super, thank you so much for helping me figure it out!

**Comment by hendrik-brower** [comment]
The libpq exposes functions to encode literals or identifiers so that one can safely construct an sql statement.  Are these functions exposed?  If not, what's it take to expose these?  Though this is not necessary for parameters, I don't believe parameters are allowed for all of these potential elements of a statement.

**Reply by isoos** [reply]
Could you please provide some examples what you are looking for?

**Reply by hendrik-brower** [reply]
Essentially the equivalent to:

32.3.4. Escaping Strings for Inclusion in SQL Commands 
char *PQescapeLiteral(PGconn *conn, const char *str, size_t length);
char *PQescapeIdentifier(PGconn *conn, const char *str, size_t length);

This would allow building an sql string with something like:

final String tbl = getUserInputForATableName();
final String val = getUserInput();
final buf = StringBuffer();
buf.write('select * from ');
buf.write(PQescapeIdentifier(tbl));
buf.write(' where x=\'');
buf.write(PQescapeLiteral(val);
buf.write('\'');
final sql = buf.toString();

In this case, the "val" would be better applied as a parameter, though there are certainly cases where it makes more sense to embed a literal within a dynamically constructed sql string.

**Reply by isoos** [reply]
I can see the appeal, but I don't have a pressing need for it right now. Would you be interested in contributing such implementation (+ test) to the package? I can review and maintain it once it is merged though, if enough test and official references are provided.
